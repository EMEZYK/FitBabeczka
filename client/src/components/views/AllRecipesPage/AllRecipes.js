import React from "react";
import { useState, useEffect } from "react";
import AllDishes from "../../ui/AllDishes/AllDishes";
import { FlexWrapper } from "../../global-styles/Flex.styled";
import CategoriesWrapper from "../../ui/CategoriesWrapper/CategoriesWrapper";
import SearchInputComponent from "../../ui/SearchInput/SearchInput";
import FooterComponent from "../../ui/Footer/Footer";
import axios from "axios";
import { Pagination } from "../../ui/Pagination/Pagination";

const AllRecipesPage = ({
  categories,
  categoriesLoadingError,
  isEditable,
  openModal,
  isAdminPanel,
}) => {
  const [recipes, setRecipes] = useState([]);
  const [searchTerm, setSearchTerm] = useState("");
  const [termToLookup, setTermToLookup] = useState("");
  const [loading, setLoading] = useState(true);

  const [limitOfRecipes] = useState(8);
  const [pageNumber, setPageNumber] = useState(0);
  const [numberOfPages, setNumberOfPages] = useState(0);
  const [numberOfRecipes, setNumberOfRecipes] = useState();

  const [deletedRecipeId, setDeletedRecipeId] = useState("");
  const [editedRecipeId, setEditedRecipeId] = useState("");

  const fetchData = async () => {
    await axios({
      url: `/recipes?limit=${limitOfRecipes}&page=${pageNumber}`,
    })
      .then((response) => {
        const limitedRecipes = response.data.filteredData;
        const recipesLength = response.data.pagination.total;
        const totalPages = response.data.pagination.pages;

        setRecipes(limitedRecipes);
        setNumberOfRecipes(recipesLength);
        setNumberOfPages(totalPages);
      })
      .catch((error) => console.error(`Error: ${error} `));
    setLoading(false);
  };

  useEffect(() => {
    if (openModal) {
      return;
    }
    fetchData();
  }, [
    limitOfRecipes,
    searchTerm === "",
    pageNumber,
    numberOfRecipes,
    deletedRecipeId,
    editedRecipeId,
    openModal,
  ]);

  const fetchSearchData = async () => {
    await axios({
      url: "/recipes/search/" + termToLookup,
    })
      .then((response) => {
        const allRecipes = response.data;
        setRecipes(allRecipes);
      })
      .catch((error) => console.error(`Error: ${error} `));
    setLoading(false);
  };

  useEffect(() => {
    fetchSearchData();
  }, [termToLookup]);

  const displayPagination = () => {
    if (termToLookup.length === 0) {
      return (
        <Pagination
          totalRecipes={numberOfRecipes}
          recipesPerPage={limitOfRecipes}
          paginate={setPageNumber}
          numberOfPages={numberOfPages}
          thisPage={pageNumber}
          setPageNumber={setPageNumber}
          pageNumber={pageNumber}
        />
      );
    }
  };

  const renderRecipes = () => {
    return (
      <AllDishes
        recipes={recipes}
        loading={loading}
        isEditable={isEditable}
        setDeletedRecipeId={setDeletedRecipeId}
        setEditedRecipeId={setEditedRecipeId}
        categories={categories}
      />
    );
  };

  return (
    <FlexWrapper height="auto" mobilePadding="0">
      <CategoriesWrapper
        categories={categories}
        categoriesLoadingError={categoriesLoadingError}
        isAdminPanel={isAdminPanel}
      />
      <SearchInputComponent
        setSearchTerm={setSearchTerm}
        setTermToLookup={setTermToLookup}
        searchTerm={searchTerm}
      />

      {renderRecipes()}
      {displayPagination()}

      <FooterComponent />
    </FlexWrapper>
  );
};
export default AllRecipesPage;
