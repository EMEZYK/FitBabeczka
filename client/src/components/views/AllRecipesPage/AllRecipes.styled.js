import styled from "styled-components";

export const PaginatedList = styled.div`
  display: flex;
  height: 3rem;
  align-items: center;
  margin: 0 0 1rem 0;
`;

export const NextPrevButton = styled.button`
  width: 2rem;
  padding: 0.3rem;
  cursor: pointer;
  background-color: #fff;
  border: 1px solid #ddd;

  a {
    text-decoration: none;
    color: #000;
  }
`;
