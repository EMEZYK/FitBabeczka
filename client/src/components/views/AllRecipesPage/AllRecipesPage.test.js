import React from "react";
import { render, cleanup, waitFor } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import AllRecipesPage from "./AllRecipes";
import "@testing-library/jest-dom";

describe("Test AllRecipesPage", () => {
  afterEach(cleanup);

  test("Should take a snapshot',", () => {
    const { asFragment } = render(<AllRecipesPage />);
    expect(asFragment(<AllRecipesPage />)).toMatchSnapshot();
  });

  test("Check if pagination list is displayed", async () => {
    render(<AllRecipesPage />);
    await waitFor(() => {
      const pagination = screen.queryByTestId("pagination");
      expect(pagination).toBeTruthy();
    });
  });

  test("Check if input search is displayed", () => {
    render(<AllRecipesPage />);

    const inputSearch = screen.getByLabelText("Search");
    expect(inputSearch).toBeInTheDocument();
  });

  test("Check if footer is displayed", () => {
    render(<AllRecipesPage />);

    const footer = screen.queryByTestId("footer");
    expect(footer).toBeInTheDocument();
  });
});
