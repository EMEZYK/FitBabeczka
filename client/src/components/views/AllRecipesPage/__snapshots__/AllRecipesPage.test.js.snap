// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AllRecipesPage Should take a snapshot', 1`] = `
<DocumentFragment>
  <div
    class="sc-papXJ dMLqgr"
    height="auto"
  >
    <div
      class="sc-iTONeN eoontZ"
    >
      <input
        aria-label="Search"
        class="sc-iAvgwm kymTGl"
        name="searchTerm"
        placeholder="Szukaj"
        type="search"
      />
      <button
        class="sc-dkzDqf sc-cTQhss gnOxCg eiqqIz"
      >
        <svg
          alt="input-loop"
          class="sc-efBctP khDFKk"
          fill="none"
          viewBox="0 0 50 50"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M11.1876 11.1875C9.89676 12.4611 8.87061 13.9776 8.16823 15.6494C7.46585 17.3213 7.10112 19.1156 7.09505 20.929C7.08897 22.7425 7.44167 24.5392 8.13283 26.2157C8.82399 27.8923 9.83997 29.4155 11.1223 30.6978C12.4045 31.9801 13.9278 32.9961 15.6044 33.6872C17.2809 34.3784 19.0776 34.7311 20.891 34.725C22.7045 34.7189 24.4988 34.3542 26.1706 33.6518C27.8425 32.9495 29.3589 31.9233 30.6326 30.6325C33.1827 28.048 34.6071 24.5598 34.5949 20.929C34.5827 17.2983 33.135 13.8197 30.5677 11.2524C28.0003 8.68504 24.5218 7.23734 20.891 7.22517C17.2603 7.21301 13.7721 8.63737 11.1876 11.1875V11.1875ZM27.0976 27.0975C25.4566 28.7385 23.2309 29.6604 20.9101 29.6604C18.5894 29.6604 16.3636 28.7385 14.7226 27.0975C13.0816 25.4564 12.1597 23.2307 12.1597 20.91C12.1597 18.5892 13.0816 16.3635 14.7226 14.7225C16.3636 13.0814 18.5894 12.1595 20.9101 12.1595C23.2309 12.1595 25.4566 13.0814 27.0976 14.7225C28.7386 16.3635 29.6606 18.5892 29.6606 20.91C29.6606 23.2307 28.7386 25.4564 27.0976 27.0975V27.0975Z"
            fill="#DEA8A8"
            fill-rule="evenodd"
          />
          <path
            d="M20.1937 33.2849C19.8453 32.9366 19.569 32.5231 19.3805 32.068C19.192 31.6128 19.095 31.125 19.095 30.6324C19.095 30.1398 19.192 29.652 19.3805 29.1969C19.569 28.7418 19.8453 28.3283 20.1937 27.9799C20.542 27.6316 20.9555 27.3553 21.4106 27.1668C21.8658 26.9783 22.3536 26.8812 22.8462 26.8812C23.3388 26.8812 23.8266 26.9783 24.2817 27.1668C24.7368 27.3553 25.1503 27.6316 25.4987 27.9799L34.3362 36.8199C34.6943 37.1659 34.98 37.5796 35.1766 38.0372C35.3731 38.4947 35.4765 38.9868 35.4809 39.4847C35.4852 39.9826 35.3903 40.4764 35.2018 40.9373C35.0132 41.3981 34.7347 41.8168 34.3827 42.1689C34.0306 42.521 33.6119 42.7995 33.151 42.988C32.6901 43.1766 32.1963 43.2714 31.6984 43.2671C31.2005 43.2628 30.7084 43.1593 30.2509 42.9628C29.7934 42.7663 29.3796 42.4806 29.0337 42.1224L20.1937 33.2849Z"
            fill="#DEA8A8"
          />
        </svg>
      </button>
    </div>
    <h2>
      Loading...
    </h2>
    <div
      class="sc-cOFTSb docskf"
      data-testid="pagination"
    >
      <button
        class="sc-BeQoi glvcdG"
      >
        <a
          href="#"
        >
          «
        </a>
      </button>
      <ul
        class="sc-dPyBCJ zOVpa"
      />
      <button
        class="sc-BeQoi glvcdG"
      >
        <a
          href="#"
        >
          »
        </a>
      </button>
    </div>
    <footer
      class="sc-jOrMOR eaxFPJ"
      data-testid="footer"
    >
      <p>
        ©2022 Ewelina Mężyk, Wszelkie prawa zastrzeżone
      </p>
    </footer>
  </div>
</DocumentFragment>
`;
